src/course-outline/drag-helper/DraggableList.jsx

```diff
@@ -0,0 +1,345 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import {
+  DndContext,
+  closestCorners,
+  KeyboardSensor,
+  PointerSensor,
+  useSensor,
+  useSensors,
+} from '@dnd-kit/core';
+import {
+  arrayMove,
+  sortableKeyboardCoordinates,
+} from '@dnd-kit/sortable';
+import { restrictToVerticalAxis } from '@dnd-kit/modifiers';
+
+import DragContextProvider from './DragContextProvider';
+import { COURSE_BLOCK_NAMES } from '../constants';
+import {
+  moveSubsectionOver,
+  moveUnitOver,
+  moveSubsection,
+  moveUnit,
+  dragHelpers,
+} from './utils';
+
+const DraggableList = ({
+  items,
+  setSections,
+  restoreSectionList,
+  handleSectionDragAndDrop,
+  handleSubsectionDragAndDrop,
+  handleUnitDragAndDrop,
+  children,
+}) => {
+  const prevContainerInfo = React.useRef();
+  const sensors = useSensors(
+    useSensor(PointerSensor),
+    useSensor(KeyboardSensor, {
+      coordinateGetter: sortableKeyboardCoordinates,
+    }),
+  );
+  const [activeId, setActiveId] = React.useState();
+  const [currentOverId, setCurrentOverId] = React.useState();
+
+  const findItemInfo = (id) => {
+    // search id in sections
+    const sectionIndex = items.findIndex((section) => section.id === id);
+    if (sectionIndex !== -1) {
+      return {
+        index: sectionIndex,
+        item: items[sectionIndex],
+        category: COURSE_BLOCK_NAMES.chapter.id,
+        parent: 'root',
+      };
+    }
+
+    // search id in subsections
+    for (let index = 0; index < items.length; index++) {
+      const section = items[index];
+      const subsectionIndex = section.childInfo.children.findIndex((subsection) => subsection.id === id);
+      if (subsectionIndex !== -1) {
+        return {
+          index: subsectionIndex,
+          item: section.childInfo.children[subsectionIndex],
+          category: COURSE_BLOCK_NAMES.sequential.id,
+          parentIndex: index,
+          parent: section,
+        };
+      }
+    }
+
+    // search id in units
+    for (let index = 0; index < items.length; index++) {
+      const section = items[index];
+      for (let subIndex = 0; subIndex < section.childInfo.children.length; subIndex++) {
+        const subsection = section.childInfo.children[subIndex];
+        const unitIndex = subsection.childInfo.children.findIndex((unit) => unit.id === id);
+        if (unitIndex !== -1) {
+          return {
+            index: unitIndex,
+            item: subsection.childInfo.children[unitIndex],
+            category: COURSE_BLOCK_NAMES.vertical.id,
+            parentIndex: subIndex,
+            parent: subsection,
+            grandParentIndex: index,
+            grandParent: section,
+          };
+        }
+      }
+    }
+    return null;
+  };
+
+  /* istanbul ignore next */
```

Comment:
@arbrandes Ahh, I meant to write a comment about it, but then forgot. I have ignored two functions from coverage, i.e., unitDragOver and subsectionDragOver. For some weird reason, dragEnd handler is not being triggered by dnd-kit while testing them.

I had already spent quite some time testing the functions related to moving/dragging blocks which doesn't work with jsdom easily as it does not implement `getClientBoundingRect` and renders all elements on top of each other at (0, 0) position. So I had to mock `collisionDetection` function which helped me test movement of blocks inside its own parent as well as across parents.

The main functions responsible to move units & subsections across parents are tested, just these functions which determine the new index and parent are ignored. 

I think now I can try to unit test them instead of depending on dnd-kit. Will update this soon.

Reply:
Thanks, it makes sense.

I still can't find the explanation in the code, though.  Did you add it?  If not, I would suggest adding it at the first occurrence of the `istanbul ignore next` of each file where this occurs.  (I expect it's just this file.)

Reply:
@arbrandes Done.